File: FrontEnd/domain/entities/Room.ts
// Room domain entities

export type RoomStatus = 'available' | 'occupied' | 'housekeeping' | 'maintenance';

export interface RoomType {
  id: string;
  tenantId: string;
  name: string;
  code: string;
  price: number;
  amenities: string[];
  imageUrl?: string;
  createdAt: string;
  updatedAt: string;
}

export interface Room {
  id: string;
  tenantId: string;
  roomTypeId: string;
  roomNumber: string;
  floor?: number;
  status: RoomStatus;
  createdAt: string;
  updatedAt: string;
  // Populated on read
  roomTypeName?: string;
}

File: BackEnd/app/models/room.py
import uuid
from datetime import datetime

from sqlalchemy import ForeignKey, Integer, String, TIMESTAMP, UniqueConstraint
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db.base import Base


class Room(Base):
    __tablename__ = "rooms"
    __table_args__ = (
        UniqueConstraint("tenant_id", "room_number", name="uq_rooms_room_number"),
    )

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    tenant_id: Mapped[uuid.UUID] = mapped_column(
        ForeignKey("tenants.id", ondelete="CASCADE"), nullable=False
    )
    room_type_id: Mapped[uuid.UUID] = mapped_column(
        ForeignKey("room_types.id", ondelete="RESTRICT"), nullable=False
    )
    room_number: Mapped[str] = mapped_column(String(20), nullable=False)
    floor: Mapped[int | None] = mapped_column(Integer)
    status: Mapped[str] = mapped_column(String(40), default="available", nullable=False)
    created_at: Mapped[datetime] = mapped_column(TIMESTAMP, default=datetime.utcnow)
    updated_at: Mapped[datetime] = mapped_column(
        TIMESTAMP, default=datetime.utcnow, onupdate=datetime.utcnow
    )

    tenant = relationship("Tenant", back_populates="rooms")
    room_type = relationship("RoomType", back_populates="rooms")
    check_ins = relationship("CheckIn", back_populates="room")
    incidents = relationship("Incident", back_populates="room")
